File: project_euler/src/problem_013.erl
%% Work out the first ten digits of the sum of the one-hundred 50-digit numbers in "problem_013.dat".

-module(problem_013).
-export([get_check_data/0, prepare_data/2, solve/1]).

-behaviour(numerical_task_behaviour).

get_check_data() ->
    [{{"problem_013.dat", 10}, "5537376230"}].

prepare_data(ModuleSourceDir, {Filename, DigitsCount}) ->
    NumberStrings = load_utils:read_strings(filename:join(ModuleSourceDir, Filename)),
    Numbers = lists:map(fun list_to_integer/1, NumberStrings),
    {Numbers, DigitsCount}.

solve({Numbers, DigitsCount}) ->
    Sum = lists:sum(Numbers),
    SumString = integer_to_list(Sum),
    lists:sublist(SumString, DigitsCount).

File: projecteuler.fsharp/ProjectEulerTasks/Problem013.fs
namespace ProjectEulerTasks

open NUnit.Framework
open ProjectEulerTasks.Utils
open System.IO

// Work out the first ten digits of the sum of the one-hundred 50-digit numbers in "problem_013.dat".

[<TestFixture>]
type Problem013() =

    let solveImpl (dataFilename: string) (count: int) =
        let sum = File.ReadLines(Path.Combine("Data", dataFilename)) |> Seq.map (fun line -> line |> bigint.Parse) |> Seq.sum
        (sum |> string).Substring(0, count)

    [<TestCase("problem_013.dat", 10, "5537376230", TimeThresholds.HardTimeLimit)>]
    member public this.Solve(dataFilename: string, count: int, expectedAnswer: string, timeLimit: int) =
        SolutionUtils.CheckSolution(timeLimit, expectedAnswer, solveImpl, dataFilename, count)
